proposals/0000-backpack.rst

```diff
@@ -0,0 +1,1185 @@
+a Haskell entity ``n`` defined in a module with the module identity
+``M`` (notice that the unique name of a declaration in a module is
+not well-specified without extra information, since only module
+names are specified in syntax) is ascribed the unique name ``M.n``:
+two identifiers are only equal if their unique names are equal.
+
+Example: the module identifier for ``Str`` from ``str-bytestring``
+is ``str-bytestring-0.2-xxx:Str``; the module identifier for
+``Concat`` frmo an uninstantiated ``concat-indef`` is
+``concat-indef-0.1-abcdefg[Str=<Str>]:Concat``.
+
+.. _ModuleSubst:
+
+A **module substitution** is a mapping from module names to modules identifiers.
+Any module name occurs only once in a substitution, and a substitution
+is in canonical form if it is sorted by module name.  We will use
+the metavariable ``S`` to represent substitutions.
+
+Module substitutions can be applied to identifiers::
+
+    -- Substitution on UnitId
+    (p[S])⟦S'⟧   = p[S⟦S'⟧]
+
+    -- Substitution on Module
+    <m>⟦⟧        = <m>
+    <m>⟦m=M,  S⟧ = M
+    <m>⟦m'=M, S⟧ = <m>⟦S⟧    (m ≠ m')
+    (P:m)⟦S⟧     = P⟦S⟧:m
+
+    -- Substitution on ModuleSubst (NOT substitution composition)
+    (m=M, S')⟦S⟧ = m=M⟦S⟧, S'⟦S⟧
+
+Mixed libraries
+---------------
+
+Libraries defines a collection of modules and dependencies parametrized
+over a set of required signatures.  Mixed libraries are an *explicit*
+variant of the libraries users write in Cabal files, where the
+instantiation of all dependencies is recorded explicitly via unit
+identities (similar to ML's *applicative functors*).  Mixed libraries
+are the output of mix-in linking performed by Cabal.
+
+Although users are not expected to ever write a mixed library directly,
+mixed libraries serve as an important intermediate language between
+GHC and Cabal: a mixed library represents a set of command
+line flags and input files that can be directly typechecked by GHC.
+
+Mixed library structure
+~~~~~~~~~~~~~~
+
+To discuss mixed libraries in a more user friendly form, we
+define an abstract syntax tree for mixed libraries and show how to translate
+this AST into the command line arguments that the compiler accepts.
+
+::
+
+    mlib  ::= "library" ComponentId "where" "{"
+                mdecl_0 ";" ... ";" mdecl_n
+              "}"
+    mdecl ::= "dependency" UnitId ModuleRenaming
+            | "module"    ModuleName
+            | "signature" ModuleName
+
+A mixed library begins with a header recording its component identity.  The
+body of a library consists of any number of dependencies, modules
+and signatures.
+
+For example, ``concat-indef`` and ``stringutils-indef`` would have the
+following ASTs::
+
+    library concat-indef-0.1-abcdefg where
+        signature Str
+        module Concat
+
+    library stringutils-indef-0.1-xxx where
+        dependency concat-indef-0.1-abcdefg[Str=<Str>]
+        module StringUtils
+
+Here is an informal translation of this AST into command line flags:
+
```

Comment:
The typechecking can be done all by itself, and the instantiation which is when we compile it. When it's fully instantiated in another library, we generate code for it.